src/oc/auth/lib/slack.clj

```diff
@@ -0,0 +1,142 @@
+(ns oc.auth.lib.slack
+  (:require [defun.core :refer (defun)]
+            [clojure.string :as s]
+            [clj-slack.oauth :as slack-oauth]
+            [clj-slack.auth :as slack-auth]
+            [clj-slack.core :as slack]
+            [clj-slack.users :as slack-users]
+            [taoensso.timbre :as timbre]
+            [oc.lib.db.common :as db-common]
+            [oc.auth.config :as config]))
+
+(def ^:private slack-endpoint "https://slack.com/api")
+(def ^:private slack-connection {:api-url slack-endpoint})
+
+(def ^:private slack
+  {:redirectURI  "/slack/auth"
+   :state        "open-company-auth"})
+
+(defn- coerce-to-user
+  "Coerce the given map to a user, return nil if any important attributes are missing"
+  [user-data]
+  (let [user-name (or (:real_name_normalized user-data)
+                      (:real_name user-data)
+                      (:name user-data))
+        split-name (when user-name (s/split user-name #"\s"))
+        name-size (count split-name)
+        splittable-name? (= name-size 2)]
+    {:user-id (db-common/unique-id)
+     :slack-id (:id user-data)
+     :name user-name
+     :first-name (cond
+                    (= name-size 1) user-name
+                    splittable-name? (first split-name)
+                    :else "")
+     :last-name (if splittable-name? (last split-name) "")
+     :avatar-url (or (:image_192 user-data)
+                     (:image_72 user-data)
+                     (:image_48 user-data)
+                     (:image_512 user-data)
+                     (:image_32 user-data)
+                     (:image_24 user-data))
+     :email (:email user-data)}))
+
+(defn get-user-info
+  [access-token scope slack-id]
+  {:pre [(string? access-token) (string? slack-id)]}
+  (let [resp (slack-users/info (merge slack-connection {:token access-token}) slack-id)]
+    (if (:ok resp)
+      (coerce-to-user (merge (:user resp) (-> resp :user :profile)))
+      (throw (ex-info "Error response from Slack API while retrieving user data"
+                      {:response resp :slack-id slack-id :scope scope})))))
+
+(defn valid-access-token?
+  "Given a Slack access token, see if it's valid by making a test call to Slack."
+  [access-token]
+  (let [conn      (merge slack-connection {:token access-token})
+        identity  (slack/slack-request conn "users.identity")
+        auth-test (slack-auth/test conn)]
+    (if (or (:ok identity) (:ok auth-test))
+      (coerce-to-user (merge (:user identity) (-> identity :user :profile)))
+      (timbre/warn "Access token could not be validated"
+                   {:identity identity :auth-test auth-test}))))
```

Comment:
`timbre/warn` return nil? or we should explicitly return an nil or false or an error message here?